src/components/molecules/VenueJoinButton/VenueJoinButton.tsx

```diff
@@ -0,0 +1,180 @@
+import React, { useCallback, useState } from "react";
+
+import { Modal } from "react-bootstrap";
+import { useForm } from "react-hook-form";
+
+import {
+  DEFAULT_PARTY_BUTTON_TEXT,
+  DEFAULT_PARTY_JOIN_MESSAGE,
+} from "settings";
+
+import { checkAccess } from "api/auth";
+
+import { AnyVenue } from "types/venues";
+
+import { useUser } from "hooks/useUser";
+
+import { getCurrentTimeInUTCSeconds, getTimeBeforeParty } from "utils/time";
+import { setLocalStorageToken } from "utils/localStorage";
+import { isTruthy } from "utils/types";
+import { joinVenue } from "utils/url";
+
+import "./VenueJoinButton.scss";
+
+export interface VenueJoinButtonProps {
+  venueId?: string;
+  venue: AnyVenue;
+  onPasswordSubmit?: () => void;
+}
+
+export interface VenueJoinButtonData {
+  password: string;
+}
+
+export const VenueJoinButton: React.FC<VenueJoinButtonProps> = ({
+  venueId,
+  venue,
+  onPasswordSubmit,
+}) => {
+  const { user } = useUser();
+  const [message, setMessage] = useState("");
+  const [isJoinVenueModalOpen, setJoinVenueModalOpen] = useState(false);
+
+  const isLoggedIn = !!user;
+  const buttonText =
+    venue.config?.landingPageConfig.joinButtonText ?? DEFAULT_PARTY_BUTTON_TEXT;
+
+  const {
+    register,
+    handleSubmit,
+    errors,
+    setError,
+    clearError,
+  } = useForm<VenueJoinButtonData>({
+    mode: "onChange",
+    reValidateMode: "onChange",
+  });
+
+  const showJoinVenueModal = useCallback(() => {
+    setJoinVenueModalOpen(true);
+  }, []);
+
+  const hideJoinVenueModal = useCallback(() => {
+    setJoinVenueModalOpen(false);
+  }, []);
+
+  const handleJoinVenue = useCallback(() => {
+    if (!venueId) return;
+
+    joinVenue(venueId, { hasVenueEntrance: !!venue.entrance?.length });
+  }, [venue, venueId]);
+
+  const changePassword = useCallback(() => {
+    setMessage("");
+    clearError();
+  }, [clearError]);
+
+  const submitPassword = useCallback(
+    async ({ password }) => {
+      if (!venueId) return;
+
+      if (!venue.access) {
+        handleJoinVenue();
+        return;
+      }
```

Comment:
Since this does more than submit the password, perhaps a better name for the `const`? Maybe `checkAccessOrJoinVenue` or similar?

Reply:
Also, should we be calling `handleJoinVenue` directly here? Or do we still want to be popping up the modal?